lib/transform-stream.js

```diff
@@ -2,30 +2,62 @@
 /* global TransformStream */
 
 /**
- * Pipe a readable stream through a transformer. Return the readable end of the
- * TransformStream.
+ * Pipe a readable stream through a transformer. Returns a result, where
+ * result.readable is the readable end of the TransformStream and
+ * result.done is a promise that fulfills or rejects once the stream is done.
  * Includes a shim for environments where TransformStream is not available.
  */
-export function transformStream (readable, transformer) {
-  // Chrome, Edge, Safari TP
+export function transformStream (sourceReadable, transformer) {
+  let transformedReadable
+  let done
+
   if (typeof TransformStream !== 'undefined') {
-    return readable.pipeThrough(new TransformStream(transformer))
+    // Chrome, Edge, Safari 14.1+
+    const transform = new TransformStream(transformer)
+
+    done = sourceReadable.pipeTo(transform.writable)
+    transformedReadable = transform.readable
+  } else {
+    // Firefox, Safari 14 and older
+    let doneCb = null
+
+    done = new Promise((resolve, reject) => {
+      doneCb = (err) => {
+        if (err) {
+          reject(err)
+        } else {
+          resolve()
+        }
+      }
+    })
+    transformedReadable = new ReadableStream(new TransformStreamSource(sourceReadable, transformer, doneCb))
   }
 
-  // Firefox, Safari 14 and older
-  return new ReadableStream(new TransformStreamSource(readable, transformer))
+  // Ensure the caller doesn't need to catch errors
+  done.catch(() => {})
+
+  return {
+    readable: transformedReadable,
+    done
+  }
 }
 
 class TransformStreamSource {
-  constructor (readable, transformer) {
+  constructor (readable, transformer, doneCb) {
     this.readable = readable
     this.transformer = transformer
+    this.doneCb = doneCb
     this.reader = readable.getReader()
   }
 
   async start (controller) {
-    if (this.transformer.start) {
-      return await this.transformer.start(controller)
+    if (this.transformer?.start) {
+      try {
+        return await this.transformer.start(controller)
```

Comment:
This was an issue before this PR, but I don't think we need the promise returned from this function to resolve to an actual value since it is just used to indicate success or failure.

So I think you can remove the `return` keyword.

Reply:
Good point!